### PR TITLE
Update Congressional Districts shape files 

### LINKED ISSUE #5663: Research using geodatabase 118th congressional district file
**What we're after:**
_Research updating the election maps with the provided [geodatabase 118th congressional district](https://www.census.gov/geographies/mapping-files/time-series/geo/tiger-geodatabase-file.html) file or [individual state files](https://www2.census.gov/geo/tiger/TIGER_RD18/LAYER/CD/)._

### Related issues
- #5655 

### Completion criteria

- [x] Research/experimentation is complete and recommended next steps are documented and shared with the team

### Tech steps or considerations

We can try the following to experiment, per conversation with @rfultz:

- Experiment with two simple neighboring states (CO and WY are probably the easiest) and examine whether they share the same values on their shared value. So if the upper left corner of both are the same, we’re out of luck, but if the Y numbers for WY’s southern border look like those for CO’s northern border, cool, the merge could probably work well—at least for 48 of them
- Experiment with the files and go through the process to get the JS/JSON files and hand-merge the new JSON then view them in the map and see if the two shapes are properly spaced or overlapping. If they’re properly spaced, we could run the script on all 56 and then merge inside json.

### PR DESCRIPTION
**Pull request by tmpayton**
## Summary (required)

- Resolves #5663

I did research and experimented with using geodatabases and the state shape files to update our Congressional district maps. I tested a few different ways, and was able to successful update our maps with each of them.  

Method 1: Merge the state shape files using [Geopandas](https://geopandas.org/en/stable/docs/user_guide.html) in python. 

Method 2: Use [QGIS](https://www.qgis.org/en/site/) to extract congressional district shape file from geodatabase 
- QGIS is free, open source software that allows users to manipulate geospatial information 

Method 3: Use [GDAL](https://gdal.org/index.html) to convert geodatabase to a shape file.
- GDAL is an open source translator library 

## Recommendation: 
 
- All methods work, but I thought GDAL is the fastest and easiest method. But it might be good to store the python script so that we have a solution to update our maps if we only have the state shape files. 

## Impacted areas of the application

General components of the application that this PR will affect:

-  Congressional District Maps 

## How to test

Prerequisites for all methods: 

For the create_districts_topo.sh script you need the following npm packages: shapefile, ndjson-cli, topojson-simplify,  topojson-server, topojson-client 

`npm install -g shapefile`
`npm install  -g ndjson-cli`
`npm install -g topojson-simplify`
`npm install -g topojson-server`
`npm install -g topojson-client`

### Method 1: Merge State Shape files 

Prerequisites: geopandas 
`pip install geopandas `

1. Checkout this branch
2. Ensure that SOURCE_FILE is set to `districts.shp` and replace all instances of `GEOID` with `GEOID20`
3. In terminal, go to data folder in fec-cms project `cd fec-cms/fec/fec/static/js/data`
4. `python3 merge_shape_files.py`
5. Go back to root of project 
6. `npm run build-js`
7. `cd fec/`
8. `./manage.py runserver` (might need to clear cookies in browser)

Note: I got a scraping ssl error at first, the fix can be found here [Scraping: SSL: CERTIFICATE_VERIFY_FAILED](https://stackoverflow.com/questions/50236117/scraping-ssl-certificate-verify-failed-error-for-http-en-wikipedia-org)

### Method 2: Use QGIS to extract shape files from geodatabase 
Prerequisites: [QGIS](https://www.qgis.org/en/site/forusers/download.html) software (this may take awhile). 

1. Download [geodatabase](https://www.census.gov/geographies/mapping-files/time-series/geo/tiger-geodatabase-file.html) from the Census website 
2. Open QGIS Layer > Add Layer > Add Vector Layer (make sure directory is highlighted)
image
3. Find the layer that says Current_Congressional_DIstricts in the Layer pane 
image
4. Right click on layer > Save features as > ESRI shape file 
- save in ~/fec-cms/fec/fec/static/js/data/src directory and name it “districts”
image
5. In terminal, go to data folder in fec-cms project `cd fec-cms/fec/fec/static/js/data`
6. `./create_districts_topo.sh`
9. Go back to root of project
10. `npm run build-js`
11. `cd fec/`
12. `./manage.py runserver` (might need to clear cookies in browser)

### Method 3: Use GDAL to convert geodatabase to shape file 
1. `brew install gdal` (this will take awhile) 
2. Download [geodatabase](https://www.census.gov/geographies/mapping-files/time-series/geo/tiger-geodatabase-file.html) from the Census website 
3. `ogr2ogr -f "ESRI Shapefile" "{folder to put shp files}" "{name of geodatabase}.gdb"` This will generate a folder with shape files. 
4. Move all "current_congressional_districts" files to  ~/fec-cms/fec/fec/static/js/data/src. There should be a .prj file, .dbf file, .shx file, and .shp file 
5. In terminal, go to data folder in fec-cms project `cd fec-cms/fec/fec/static/js/data`
6. Open file create_districts_topo.sh
 -  ensure that the variable SOURCE_FILE is set to the name of your shape file from step 4. 
 - ensure that on line 19 & 21 `d.properties.GEOID` says GEOID and not GEOID20 
7. `./create_districts_topo.sh`
8. Go back to root of project
9. `npm run build-js`
10. `cd fec/`
11. `./manage.py runserver` (might need to clear cookies in browser)

Note: The GEOID property is different in the national shape files(GEOID) vs the individual state shapefiles (GEOID20). Depending on which method is used this needs to be changed in the ./create_districts_topo.sh script.

### CODE REVIEW

**Review by patphongs — COMMENTED** [review]
@tmpayton Really wonderful work on this! Thanks for pairing with me on testing the changes, the district maps look good and the JSON is updated with the new boundaries 🎉 Just a few small things we discussed updating, I'm noting them here. 

For reference, the changes to the numbers of districts have been documented in this [past PR](https://github.com/fecgov/fec-cms/pull/5205).

We also need to update the content in the "Information about redistricting and the 2020 Decennial Census" note.

> Following the 2020 Census, some states may redraw their districts. Some candidates may subsequently amend their registration forms to reflect new district numbers. Candidates are listed using information from their registration forms and may run in different districts than in the past. The FEC will update its map to reflect the new congressional districts and boundaries once information has been made available from each state.

**Comment by PaulClark2** [comment]
This is great work @tmpayton. Thank you for investigating and testing multiple methods.

**Comment by JonellaCulmer** [comment]
@tmpayton @patphongs We discussed holding off on merging this until the 21.2 release so that the default year on the election search will change to 2024 and this language and the updated maps make sense. 

Here is the updated language:

Updated accordion:
 Following the 2020 Census, some states have redrawn their district maps. The FEC has updated its map to reflect the new congressional districts and boundaries.

Missing maps text (This text only appears for election years prior to 2024.)
District maps are only available for elections in 2024 and future election years.

**Comment by patphongs** [comment]
@tmpayton In our constants.py, please also update the `DISTRICT_MAP_CUTOFF` to 2024. As we won't want to show the individual state district maps before that cut off year.

**Comment by patphongs** [comment]
@tmpayton Could you please also do these changes that @JonellaCulmer requested? Once this is made, we're good to merge this in, thanks!

> Here is the updated language:
> 
> Updated accordion:  Following the 2020 Census, some states have redrawn their district maps. The FEC has updated its map to reflect the new congressional districts and boundaries.
> 
> Missing maps text (This text only appears for election years prior to 2024.) District maps are only available for elections in 2024 and future election years.

**Comment by johnnyporkchops** [comment]
@tmpayton you wrote:

"All methods work, but I thought GDAL is the fastest and easiest method. But it might be good to store the python script so that we have a solution to update our maps if we only have the state shape files."
Which script are you  referring to here?

Also do you think we should commit the current zip to the repo too  so we have a stored copy of it ? I know it's a big file and the link to download is in the edited  instructions, so maybe that is sufficient, Just a thought.

File: fec/fec/static/js/data/merge_shape_files.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import shutil
```
Comment (tmpayton) [inline fec/fec/static/js/data/merge_shape_files.py:1]: @johnnyporkchops Thank you for the review! This is the python script that I am referring to.

**Comment by tmpayton** [comment]
> @tmpayton you wrote:
> 
> "All methods work, but I thought GDAL is the fastest and easiest method. But it might be good to store the python script so that we have a solution to update our maps if we only have the state shape files." Which script are you referring to here?
> 
> Also do you think we should commit the current zip to the repo too so we have a stored copy of it ? I know it's a big file and the link to download is in the edited instructions, so maybe that is sufficient, Just a thought.

I think that a link to the website should be sufficient. Looking at the Census website, they have shape files from 1992 and 2000 still available.

**Comment by tmpayton** [comment]
> @tmpayton Could you please also do these changes that @JonellaCulmer requested? Once this is made, we're good to merge this in, thanks!
> 
> > Here is the updated language:
> > Updated accordion:  > Following the 2020 Census, some states have redrawn their district maps. The FEC has updated its map to reflect the new congressional districts and boundaries.
> > Missing maps text (This text only appears for election years prior to 2024.) District maps are only available for elections in 2024 and future election years.

@patphongs All done! Thanks

**Review by patphongs — APPROVED** [review]
Changes look great! Thanks @tmpayton for your work on this!

**Pull request merged**